Implemented most float instructions

jaborn opened the pull request:
except for 'sflot'/'sflotu'/'sfloti'/'sflotui'

At some instructions one can hand over a rounding mode which partly was a bit difficult. Therefore I only work with these modi while casting a float to an integer and otherwise I drop it.

jaborn committed:
Implemented most float instructions
except for 'sflot'/'sflotu'/'sfloti'/'sflotui'

jaborn committed:
Remove tabs

jaborn committed:
Remove tabs

jonas-schievink commented on src/machine/behavior/is/float/fcmp.rs:
```diff
@@ -1,6 +1,17 @@
+        state.gpr[x] = (-1 as i64).into();
+    } else {
+        state.gpr[x] = 0i64.into();
+    }
```
You should be able to use the same trick as the integer-compare instruction uses, right @teiesti ?

teiesti replied:
That's funny... I was just writing about that. The short answer is: Not really. The long answer will follow soon...

teiesti replied:
The long answer:

f64 has the values -inf, +inf and NaN and is therefore PartialEq while u64 is Eq. Therefore, we have a method called [partial_cmp()][10] returning an `Option<Ordering>` which is not very helpful in this context.

I suggest to implement FCMP like this:

```rust
pub fn fcmp(state: &mut State, x: u8, y: u8, z: u8) {
    // Load operands
    let op1: f64 = state.gpr[y].into();
    let op2: f64 = state.gpr[z].into();

    // Execute
    let res: i64 = match op1 - op2 {
        d if d > 0.0 =>  1,
        d if d < 0.0 => -1,
        _            =>  0,
    }

    // Store
    state.gpr[x] = res.into();
}
```

[10]: https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#tymethod.partial_cmp

jonas-schievink replied:
True, that does make it a little harder. However, the only case where the result is `None` means that either operand was `NaN` (infinities can be compared just fine), in which case the result is `.unwrap_or(0)`.

teiesti replied:
So what we actually need is

```rust
let res = op1.partial_cmp(&op2).unwrap_or(std::cmp::Ordering::Equal) as i64;
```

Looks quite beautiful, but I think it needs a comment to explain why it is correct.

jonas-schievink commented on src/machine/behavior/is/float/fcmpe.rs:
```diff
@@ -1,6 +1,21 @@
+pub fn fcmpe(state: &mut State, x: u8, y: u8, z: u8) {
+    // Load operands
+    let op1: u64 = state.gpr[y].into();
+    let op2: u64 = state.gpr[z].into();
```
Shouldn't these be floats?

jaborn replied:
Of course!

jaborn committed:
Change type of operands from u64 into f64

jaborn committed:
Change type of operands

teiesti commented on src/machine/behavior/is/float/fcmpe.rs:
```diff
@@ -1,6 +1,21 @@
+    if res <= eps as i64 {
+        state.gpr[x] = 0i64.into();
+    } else if op1 > op2 {
+        state.gpr[x] = 1i64.into();
```
It would be nice if we do not combine the execution and result storing. (See my solution for FCMP).

Maybe it is also possible to implement a helper function for FCMP and FCMPE since FCMP is a specialized version of FCMPE.

teiesti commented on src/machine/behavior/is/float/feql.rs:
```diff
@@ -1,6 +1,14 @@
+        state.gpr[x] = 1i64.into();
+    } else {
+        state.gpr[x] = 0i64.into();
+    }
```
Try this:

```rust
// Execute
let res = (op1 == op2) as i64;

// Store
state.gpr[x] = res.into();
```

teiesti commented on src/machine/behavior/is/float/feqle.rs:
```diff
@@ -1,6 +1,19 @@
+    let eps: f64 = state.sr[R::E].into();
+
+    // Execute
+    let res = ((op1 as i64).wrapping_sub(op2 as i64)).abs();
```
`op1 as i64` will never ever work!

Try this:

```rust
// Execute
let gap = (op1 - op2).abs();
let res = (gap <= eps) as i64;

// Store
state.gpr[x] = res.into();
```

teiesti commented on src/machine/behavior/is/float/fint.rs:
```diff
@@ -1,6 +1,22 @@
+    // Execute
+    let mut res: f64;
+    match y {
+        1 => res = op1.round(),    // round off
```
```rust
let res: f64 = match y {
    1 => op1.round(),
    // ...
}
```

teiesti commented on src/machine/behavior/is/float/fint.rs:
```diff
@@ -1,6 +1,22 @@
+        3 => res = op1.floor(),    // round down
+        4 => {                     // round near to ten
+            res = (op1 + 5.0) / 10.0;
+            res = res.floor() * 10.0;
```
Round near does not mean round near to ten! It says: Round to the closer integer.

teiesti commented on src/machine/behavior/is/float/fint.rs:
```diff
@@ -1,6 +1,22 @@
+            res = (op1 + 5.0) / 10.0;
+            res = res.floor() * 10.0;
+        },
+        _ => res = op1.round(),    // round off
```
I don't think we should support a "default case". I think it's better to panic here!

jaborn replied:
But what if no specific rounding mode was committed? In this case it will panic too which is not what should happen, or am I wrong?

teiesti replied:
The question is: Is it legal for the assembler not to specify a rounding mode?

I mean, the assembler must put a value into y. Therefore, it can put one of {1, ..., 4} into y and any other value is illegal (== undefined behavior).

Is there a register that specifies a "default rounding mode"?

I am not really sure.

jaborn replied:
Ok I know what you mean. Sounds as if it's better to panic. As I read the declaration I thought it's also possible to call the instruction without a rounding mode but it makes no sense. I will change the default case.

LukasKalbertodt replied:
The CPU wouldn't panic, right? 😉 
I suspect that there is either some default behavior or an interrupt is triggered. No?

jonas-schievink replied:
just panic and put a fixme there, i want to see this thing actually execute something now

teiesti commented on src/machine/behavior/is/float/fix.rs:
```diff
@@ -1,6 +1,22 @@
-pub fn fix(_state: &mut State, _x: u8, _y: u8, _z: u8) {
-    unimplemented!();
-}
+pub fn fix(state: &mut State, x: u8, y: u8, z: u8) {
```
As far as I understand, FIX is not equal to FINT:
1. FINT round to an integer.
2. FIX rounds to fixed point number, but I do not really know how.

Therefore I suggest to leave this instruction unimplemented for the moment.

teiesti commented on src/machine/behavior/is/float/fixu.rs:
```diff
@@ -1,6 +1,23 @@
 
-pub fn fixu(_state: &mut State, _x: u8, _y: u8, _z: u8) {
-    unimplemented!();
+pub fn fixu(state: &mut State, x: u8, y: u8, z: u8) {
```
Same as FIX.

teiesti commented on src/machine/behavior/is/float/flot.rs:
```diff
@@ -1,6 +1,19 @@
+        2 => unimplemented!(),
+        3 => unimplemented!(),
+        4 => unimplemented!(),
+        _ => res = res,
```
I know, I suggested to implement this instruction like this but I am still not sure if this is right. Can somebody look up the technicalities in the MMIXware documents?

teiesti commented on src/machine/behavior/is/float/floti.rs:
```diff
@@ -1,6 +1,19 @@
+        2 => unimplemented!(),
+        3 => unimplemented!(),
+        4 => unimplemented!(),
+        _ => res = res,
```
Same as FLOT.

teiesti commented on src/machine/behavior/is/float/flotu.rs:
```diff
@@ -1,6 +1,19 @@
+        2 => unimplemented!(),
+        3 => unimplemented!(),
+        4 => unimplemented!(),
+        _ => res = res,
```
Same as FLOT.

teiesti commented on src/machine/behavior/is/float/flotui.rs:
```diff
@@ -1,6 +1,20 @@
+        2 => unimplemented!(),
+        3 => unimplemented!(),
+        4 => unimplemented!(),
+        _ => res = res,
```
Same as FLOT.

teiesti commented on src/machine/behavior/is/float/fsqrt.rs:
```diff
@@ -1,6 +1,21 @@
+    // Execute
+    let mut res: f64 = op1.sqrt();
 
+    // Rounding
```
I don't think we can do anything about the rounding mode here, since rounding happens within `op1.sqrt()`. I don't know what we can do to fit the requirements.

jonas-schievink replied:
Just throw `unimplemented!()`, non-standard rounding modes are almost never used

teiesti commented on src/machine/behavior/is/float/fun.rs:
```diff
@@ -1,6 +1,19 @@
+    let op1: f64 = state.gpr[y].into();
+    let op2: f64 = state.gpr[z].into();
 
+    // Execute
```
Try somethink like:

```rust
let res: i64 = if ...
```

teiesti commented on src/machine/behavior/is/float/fune.rs:
```diff
@@ -1,6 +1,19 @@
 
+    // Store result
+    if res >= eps as i64 {
+        state.gpr[x] = 1u64.into();
```
Same as FUN.

jaborn committed:
Several adaptations

jaborn committed:
Remove trailing whitespaces

jaborn committed:
Change default case

jaborn committed:
Change rounding

jaborn committed:
Enhancement

jaborn committed:
Bugfix

teiesti committed:
Probably wrong code with FIXME annotated

We need to merge this branch now and have no time to double-check these things.
